### PR TITLE
app/vmauth: reader pool to reduce gc & mem alloc

### PR DESCRIPTION
**Pull request by shichanglin5**
follow up https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6446

### CODE REVIEW

File: app/vmauth/main.go [hunk 1]
```diff
@@ -200,10 +200,11 @@ func processRequest(w http.ResponseWriter, r *http.Request, ui *UserInfo) {
 		up, hc = ui.DefaultURL, ui.HeadersConf
 		isDefault = true
 	}
+	r.Body = getReadTrackingBody(r.Body, int(r.ContentLength))
+	defer func() {
+		putReadTrackingBody(r.Body.(*readTrackingBody))
+	}()
```
Comment (f41gh7) [inline app/vmauth/main.go:206]: I think it's correct to check for the max value of ContentLength header. It prevents possible DDoS attacks.
```suggestion
        if int(r.ContentLength) < maxRequestBodySizeToRetry.IntN() {
           r.Body = getReadTrackingBody(r.Body, int(r.ContentLength))
	   defer func() {
		putReadTrackingBody(r.Body.(*readTrackingBody))
	}()
        }
```

**Review by f41gh7 — APPROVED** [review]
LGTM

File: app/vmauth/main_test.go [hunk 2]
```diff
@@ -85,6 +82,125 @@ func TestReadTrackingBodyRetryFailure(t *testing.T) {
...
+// 1. When writing data downstream, buf only caches part of the data because the downstream connection is disconnected.
+// 2. retry request: because buf caches some data, first read buf and then read stream when retrying
+// 3. retry request: the data has been read to buf in the second step. if the request fails, retry to read all buf later.
+func TestRetryReadSuccessAfterPartialRead(t *testing.T) {
+	f := func(s string) {
+		rtb := &readTrackingBody{
+			r:   io.NopCloser(bytes.NewBufferString(s)),
+			buf: make([]byte, 0, len(s)),
+		}
+		// 读取全部数据
```
Comment (zekker6) [inline app/vmauth/main_test.go:95]: Could you translate this one?

**Review by zekker6 — APPROVED** [review]
LGTM, see 1 comment

File: app/vmauth/main_test.go (hunk 2 above)
Comment (shichanglin5) [reply app/vmauth/main_test.go]: done

**Comment by f41gh7** [comment]
related issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6445

File: app/vmauth/main.go [hunk 3]
```diff
@@ -200,10 +200,13 @@ func processRequest(w http.ResponseWriter, r *http.Request, ui *UserInfo) {
 		up, hc = ui.DefaultURL, ui.HeadersConf
 		isDefault = true
 	}
-	maxAttempts := up.getBackendsCount()
-	r.Body = &readTrackingBody{
-		r: r.Body,
+	if r.ContentLength <= int64(maxRequestBodySizeToRetry.IntN()) {
```
Comment (hagen1778) [inline app/vmauth/main.go:203]: @shichanglin5 have you observed CPU usage reduction after using sync pool?
What's your typical workload for vmauth: read, write or read&write queries?

Comment (shichanglin5) [reply app/vmauth/main.go:203]: I have separated the read and write workload, where `vmauth-apm` is used for writing, and `vmauth-meta` is used for querying. I have just downgraded `vmauth` to the previous version (without using the reader pool). Below is a comparison of the monitoring data after the downgrade.

**vmauth-apm (for write request)**
image
image


**vmauth-meta (for read request)**
image
image

Comment (shichanglin5) [reply app/vmauth/main.go:203]: RSS Mem 
image

File: app/vmauth/main.go [hunk 4]
```diff
@@ -503,60 +506,70 @@ type readTrackingBody struct {
...
+	}
+	return rtb.readFromStream(p)
+}
 
+func (rtb *readTrackingBody) readFromStream(p []byte) (int, error) {
 	if rtb.r == nil {
 		return 0, fmt.Errorf("cannot read data after closing the reader")
 	}
-
 	n, err := rtb.r.Read(p)
```
Comment (f41gh7) [inline app/vmauth/main.go:544]: It'd be great to move check  for cannotRetry back, it should save CPU usage for limit check at read calls that exceeds configured limit

```go
      if rtb.cannotRetry {
		return n, err
	}
       if rtb.offset+n > maxRequestBodySizeToRetry.IntN() {
		rtb.cannotRetry = true
                return n, err
	}
```

**Comment by shichanglin5** [comment]
feel free to make changes to this PR if the code or title is not optimal

**Comment by f41gh7** [comment]
> feel free to make changes to this PR if the code or title is not optimal

 Sure, thanks for understanding. Currently we want additionally evaluate changes and possible performance impact.  vmauth is one of components for managed cloud.

**Comment by valyala** [comment]
FYI, this pull request has been reverted in the commit ad6af95183a8133c9def59a15320de6b0cd7cfd9 (see the commit description for more details on why it has been reverted). Instead, new commit has been created - 7ee57974935a662896f2de40fdf613156630617d .

**Comment by hagen1778** [comment]
@valyala @shichanglin5 usage of sync.Pool for req body was reverted in https://github.com/VictoriaMetrics/VictoriaMetrics/commit/80ead7cfa46516ce0b3775cd7dfb6b0f0ac5bea8. See commit description for more details.

**Pull request merged**